R/pkg_ref_class.R

```diff
@@ -122,69 +141,110 @@ as_pkg_ref.pkg_ref <- function(x, ...) {
 
 #' @importFrom utils installed.packages available.packages packageVersion
 #' @export
-as_pkg_ref.character <- function(x, repos = getOption("repos"), ...) {
-  ip <- memoise_installed_packages()
+as_pkg_ref.character <- function(x, repos = getOption("repos", "https://cran.rstudio.com"),
+                                 source = NULL, lib.loc = NULL, ...) {
 
-  # case when only a package name is provided
-  #   e.g. 'dplyr'
-  # regex from available:::valid_package_name_regexp
-  if (grepl("^[[:alpha:]][[:alnum:].]*[[:alnum:]]$", x)) {
+  dots <- list(...)
 
-    # if locally installed
-    if (x %in% ip[,"Package"]) {
-      path <- find.package(x)
-      version <- utils::packageVersion(x, lib.loc = dirname(path))
+  pkg_source_ <- ifelse(is.null(source),
+                        determine_pkg_source(x, source, repos),
+                        source)
+
+  stopifnot(pkg_source_ %in% c("pkg_install", "pkg_source", "pkg_cran_remote",
+                               "pkg_bioc_remote", "pkg_missing"))
+
+  switch(pkg_source_,
+         pkg_install = {
+           path <- find.package(x, lib.loc = lib.loc)
+           version <- utils::packageVersion(x, lib.loc = dirname(path))
+
+           new_pkg_ref(x,
+                       version = version,
+                       path = path,
+                       source = "pkg_install")
+         },
+         pkg_source = {
+
+           desc <- read.dcf(file.path(x, "DESCRIPTION"))
+           name <- unname(desc[,"Package"])
+
+           new_pkg_ref(name,
+                       version = package_version(desc[,"Version"][[1]]),
+                       path = normalizePath(x),
+                       source = "pkg_source")
+         },
+         pkg_cran_remote = {
+           ap <- memoise_available_packages(repos = repos)
+           info <- ap[ap[,"Package"] == x,,drop = FALSE]
+
+           new_pkg_ref(x,
+                       version = info[,"Version"],
+                       repo = info[,"Repository"],
+                       source = c("pkg_cran_remote"))
+         },
+         pkg_bioc_remote = {
+           bp <- memoise_bioc_available()
+           info <- bp[bp[,"Package"] == x,,drop = FALSE]
+
+           new_pkg_ref(x,
+                       version = info[,"Version"],
+                       repo = "https://bioconductor.org/packages/release/bioc",
+                       source = c("pkg_bioc_remote"))
+         },
+         pkg_missing = {
+           new_pkg_ref(x,
+                       source = c("pkg_missing"))
+         })
+}
 
-      return(new_pkg_ref(x,
-        version = version,
-        path = path,
-        source = "pkg_install"))
+#' Determine the intended source of a new package
+#'
+#' @param x Package name or path to package
+#' @param source type of source passed in `pkg_ref`
+#' @return one of c('pkg_install', 'pkg_install', 'pkg_cran_remote', 'pkg_bioc_remote')
```

Comment:
will `pkg_missing` also be one of return?

Reply:
Updated